tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor.Tests/RazorHelperUpdaterTests.cs

```diff
@@ -0,0 +1,193 @@
+            var updater = mock.Create<RazorHelperUpdater>();
+            var inputs = await GetRazorCodeDocumentsAsync(mock).ConfigureAwait(true);
+
+            // Act
+            var result = await updater.IsApplicableAsync(context.Object, inputs, CancellationToken.None).ConfigureAwait(true);
+
+            // Assert
+            var fileUpdaterResult = Assert.IsType<FileUpdaterResult>(result);
+            Assert.Equal(expectedResult.Result, fileUpdaterResult.Result);
+            Assert.Collection(fileUpdaterResult.FilePaths, expectedResult.FilePaths.Select<string, Action<string>>(e => a => Assert.EndsWith(e, a, StringComparison.Ordinal)).ToArray());
+        }
+
+        [Fact]
+        public async Task ApplyNegativeTests()
+        {
+            using var mock = await GetMock(null).ConfigureAwait(true);
+            var updater = mock.Create<RazorHelperUpdater>();
+
+            await Assert.ThrowsAsync<ArgumentNullException>("context", () => updater.ApplyAsync(null!, Enumerable.Empty<RazorCodeDocument>().ToImmutableArray(), CancellationToken.None))
+                .ConfigureAwait(true);
+        }
+
+        [Theory]
+        [MemberData(nameof(HelperTestViews))]
+        public async Task ApplyTests(string projectPath, FileUpdaterResult expectedResult)
+        {
+            // Arrange
+            using var mock = await GetMock(projectPath).ConfigureAwait(true);
+            var context = mock.Mock<IUpgradeContext>().Object;
+            var updater = mock.Create<RazorHelperUpdater>();
+            var inputs = await GetRazorCodeDocumentsAsync(mock).ConfigureAwait(true);
+
+            // Act
+            var result = await updater.ApplyAsync(context, inputs, CancellationToken.None).ConfigureAwait(true);
+
+            // Assert
+            var fileUpdaterResult = Assert.IsType<FileUpdaterResult>(result);
+            Assert.True(fileUpdaterResult.Result);
+            Assert.Collection(fileUpdaterResult.FilePaths, expectedResult.FilePaths.Select<string, Action<string>>(e => a => Assert.EndsWith(e, a, StringComparison.Ordinal)).ToArray());
+
+            // Confirm that files are updated as expected
+            var projectFiles = context.CurrentProject!.FileInfo.Directory!.GetFiles("*.*", new EnumerationOptions { RecurseSubdirectories = true });
+            var expectedFiles = new DirectoryInfo($"{Path.GetDirectoryName(projectPath)}.Fixed").GetFiles("*.*", new EnumerationOptions { RecurseSubdirectories = true });
+            Assert.Collection(projectFiles, expectedFiles.Select<FileInfo, Action<FileInfo>>(e => a =>
+            {
+                Assert.Equal(File.ReadAllText(e.FullName), File.ReadAllText(a.FullName));
+            }).ToArray());
+        }
+
+        public static IEnumerable<object[]> HelperTestViews =>
+            new List<object[]>
+            {
+                new object[]
+                {
+                    "RazorHelperUpdaterViews/OneHelper/Test.csproj",
```

Comment:
Can we mock this stuff without actually having full on projects? Something similar to the Microsoft.CodeAnalysis.Testing?

Reply:
That's a good question. The challenge will be that we need RazorCodeDocuments (and the generated C# that the Razor engine produces for them) and I'm not sure whether we can get that from any existing test helpers. We'll probably need to write our own framework that takes text for before and after razor docs and checks that code fixes apply as expected. I'm sure we could do that, but it would be a bit of an investment. I'll think about whether it's something I want to add now or later.